Added leave game option, which deletes game if empty

ImmortalMagi opened the pull request:
I have aded an option for a player to leave a game. If there are no
players, then the game is destroyed.

I should have done this with test driven development, but I didn't. I noticed the rails app seems to be using rspec for testing. Also need to run rubocop.

Commit:
Added leave game option, which deletes game if empty

I have aded an option for a player to leave a game. If there are no
players, then the game is destroyed

Commit:
Added field to Players 'Arthur'

There is a new attribute for players, Arthur. This is a boolean and is
by default false. Need to add setting to make the first player to join a
game Arthur.

Commit:
Created a new method to set the first player to Arthur

Need to still display them as Arthur in game

david-mears-2 commented on app/controllers/players_controller.rb:
```diff
@@ -9,6 +9,7 @@ def new
   def create
     @player = Player.new(player_params)
     @player.game_id = @game.id
+    if no_arthur? then @player.arthur = true end
```
a shorter one-line syntax for conditionals: `@player.arthur = true if no_arthur?`

david-mears-2 replied:
So whoever joins the game first - i.e. whoever creates the game - has to be Arthur?

Perhaps it would be better to assign Arthurness when the game is quorate.

david-mears-2 replied:
That would also allow us to write the assignment as a callback on the model (to run whenever a player is associated with a game), again keeping the controller 'thin' - not that I've been especially careful about a thin controller myself.

ImmortalMagi replied:
Yeah, that sounds plausible (particularly moving it to the model). I thought that having it given to the first player means that there is always a player with Arthur, but at the moment there is no way to assign it. The first player could also leave and a new player would have to join to get a new Arthur.

david-mears-2 replied:
At the moment, I am having the game-creator specify the number of players in advance, with 3 minimum. I wonder if I should validate 3 minimum in the database, in the form model, or in the form field? Form model feels like a happy medium. Where would you do such a validation if doing a Python app?

david-mears-2 commented on app/controllers/players_controller.rb:
```diff
@@ -33,4 +33,13 @@ def check_if_game_is_full
     redirect_to root_path, alert: "Sorry, the game ‘#{@game.slug}’ is full."
   end
+
+  def no_arthur?
```
Can you put this logic into the model, to help us keep logic out of the controller?

david-mears-2 replied:
'thin controllers' is a RoR catchphrase

ImmortalMagi replied:
I had never head that before, I assumed generic code like this went into the controller, I'll move it into the model

david-mears-2 commented on app/views/games/show.html.erb:
```diff
@@ -46,3 +46,4 @@
   <% end %>
 </section>
 
+<%= button_to "Leave Game", :action => "leave_game", :slug => @game.slug %>
```
rubocop might say use this syntax: `action: "leave_game"`

david-mears-2 commented on app/controllers/games_controller.rb:
```diff
@@ -33,6 +33,13 @@ def games_not_found
+  def leave_game
+    @game = Game.find_by(slug: slug_param) or games_not_found
+    @game.players.delete(session_player)
+    if @game.players.empty? then @game.destroy end
```
Another opportunity for a callback on the model.

david-mears-2 replied:
This might be too much complication at this stage, but if instead of destroying the game you were to mark the game with a status 'trashed', that would allow us to give a message 'That game has ended' on website.com/game/123/show, as opposed to 'That game doesn't exist'. Only if you feel like it.

david-mears replied:
In the lobby-page-2 branch I have added a migration for 'status', so once that is merged, we can do `game.status = 'trashed'` (and update set_game to account for this - shouldn't be returning any trashed games).

david-mears-2 reviewed (commented):
This is very welcome, and requires a test.

Commit:
created new method has_arthur? in game models

Commit:
added before action set game, changed players controler to use games model method

david-mears-2 commented on app/controllers/games_controller.rb:
```diff
@@ -33,6 +31,12 @@ def games_not_found
   end
 
+  def leave_game
+    @game.players.delete(session_player)
```
Prefer `destroy` to `delete` because it calls the callbacks.

david-mears-2 commented on app/controllers/games_controller.rb:
```diff
@@ -61,4 +65,8 @@ def set_player
   end
+
+  def set_game
+    @game = Game.find_by(slug: slug_param) or return redirect_to games_not_found_path(slug: slug_param)
```
I *think* we can drop the `return` here now, because we are on the last line of a method, so it will be returned anyway.

david-mears-2 commented on app/views/games/show.html.erb:
```diff
@@ -40,3 +40,4 @@
   <% end %>
 </section>
 
+<%= button_to "Leave Game", action: "leave_game", slug: @game.slug %>
```
Not sure if this should be part of this PR or if you want to move on to something else, but if we are doing this nice clear-up stuff after clicking 'leave game', it would be nice to ensure the same happens if they leave the game by another method, e.g. by clicking back in the browser. I'll put it on the todos - unless you fancy doing a little JS that alerts the player when they try clicking back?

david-mears-2 commented on config/routes.rb:
```diff
@@ -7,5 +7,7 @@
     resources :players, only: %i[create new]
   end
+
+  post '/games/leave_game', to: 'games#leave_game'
```
It would be nice to have the slug as part of the route if possible.

`post '/games/(/:slug)/players/(/:id)/leave` would match the established pattern: see '[nested resources](https://guides.rubyonrails.org/routing.html#nested-resources)'.

So:

```ruby
resources :games, only: %i[create new show], param: :slug do
  resources :players, only: %i[create new] do
    post '/leave_game', to: 'games#leave_game'
  end
end
```

For example. Or you can scope it to game (not game and player) if you like.

david-mears commented:
> This is very welcome, and requires a test.

We need to have a click of 'leave game' in the feature tests.

david-mears-2 commented on app/models/game.rb:
```diff
@@ -6,6 +6,13 @@ class Game < ApplicationRecord
 
+  # rubocop:disable Naming/PredicateName
+  def has_arthur?
+    players.each { |player| return true if player.arthur? }
```
I found a slightly more readable syntax:
```ruby
players.any?(&:arthur?)
```

david-mears-2 replied:
Which might arguably even negate the need for a method, since it's so short...

david-mears-2 committed:
Redirect to a trashed game page if the game is trashed

Using the status 'trashed' rather than destroying the game object.

david-mears-2 committed:
tidying

david-mears-2 committed:
Remove unnecessary has_arthur? method and specs

david-mears-2 committed:
Add before action for show game: destroy_session_player_if_from_another_game

david-mears-2 committed:
Replace before action:

Neater to do check_if_session_player_belongs_to_game than
destroy_session_player_if_from_another_game
